rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,337 @@
+| -------- | ---------- | ----------- | ------------------------------------------
+| `from`   | `Void`     | `Bytes`     | Loads bytes from a source as a side effect
+| `to`     | `Bytes`    | `Void`      | Writes bytes to a sink as a side effect
+| `read`   | `Bytes`    | `Arrow`     | Parses a specific format into Arrow
+| `write`  | `Arrow`    | `Bytes`     | Prints input in a specific format
+| `pull`   | `Void`     | `Arrow`     | Loads Arrow from a remote source
+| `push`   | `Arrow`    | `Void`      | Stores Arrow in a remote source
+
+#### Computation
+
+The following operators have a fixed input and output type of `Arrow`. We can
+group them into "filter" operators that do not change the schema, and "reshape"
+operators where execution yields a new schema.
+
+##### Filter
+
+| Operator | Non-Blocking | Description
+| -------- |:------------:| ------------------------------------------
+| `head`   | ✅           | Retrieve first N records
+| `tail`   | ❌           | Retrieve last N records
+| `where`  | ✅           | Filter with expression
+| `sort`   | ❌           | Sort according to field
+| `uniq`   | ✅           | Produce deduplicated output
+
+##### Reshape
+
+| Operator    | Non-Blocking | Description
+| ----------- |:------------:| ------------------------------------------
+| `put`       | ✅           | Select a set of columns (projection)
+| `drop`      | ❌           | Remove a set of columns (projection)
+| `replace`   | ✅           | Replaces field described by extractors with values
+| `extend`    | ✅           | Adds new fields with (initially fixed) values
+| `rename`    | ✅           | Renames schema meta data (type & fields)
+| `summarize` | ❌/✅[^1]    | Aggregate group and compute a summary function
+
+[^1]: Depending on the aggregate function.
+
+### Pipeline Execution
+
+Let's take a step back and just assume that we add a new `exec` command that
+just executes a pipeline.
+
+Then we can model the ingestion as follows: load data via stdin, push into a
+parser, then send off to a remote VAST node. For example, we would rewrite
+`vast import zeek` as:
+
+```bash
+vast exec 'from - | read zeek | push vast://1.2.3.4'
+```
+
+Likewise, `vast export json EXPRESSION` would become:
+
+```bash
+vast exec 'pull vast://1.2.3.4 |
+           where EXPRESSION |
+           write json |
+           to -'
+```
+
+The beauty is that VAST nodes are now longer required. Pipelines provide value
+locally as well:
+
+```bash
+vast exec 'from - |
+           where EXPRESSION |
+           write json |
+           to -'
+```
+
+What would the role of the node be then? Basically a **pipeline manager** with
+persistent state. A lot of things wouldn't change, e.g., starting a node:
+
+```bash
+vast -e 1.2.3.4 start
+```
+
+There are of course convenience defaults we can use, like assuming `from -` and
+`to -` being the default. This would make the above look like this:
+
+```bash
+vast exec 'where EXPRESSION |
+           write json
+```
```

Comment:
This doesn't make sense to me. `where` takes Arrow as input, but `from -` has Bytes as output.

Reply:
Good catch. I added a `read zeek` operator to fix it.